Add a clang-format check to GH actions

mjschmidt271 opened the pull request:
Adds a github workflow, to run on PRs, that determines adherence to clang-format style requirement.

---

This new GitHub Action runs as a check on PRs. It is intended to do a formatting check **only on the files changed by the PR** using the `clang-format` tool. Note, however, that `clang-format` will be applied to the **entire file**, and not only the lines that were changed.

For this workflow, I compared a couple of the most used Actions from the GitHub Marketplace, and thought that the diff provided by [`DoozyX/clang-format-lint-action`](https://github.com/DoozyX/clang-format-lint-action) added useful information (See [this output](https://github.com/E3SM-Project/E3SM/actions/runs/14764937805/job/41454174819) for an example.), as opposed to only the `clang-format` error output, as provided by [`jidicula/clang-format-action`](https://github.com/jidicula/clang-format-action) ([example output](https://github.com/E3SM-Project/E3SM/actions/runs/14743092536/job/41385214524)).

- [x] ~~I do like that the latter provides a summary on the page that denotes which files fail the `clang-format`, but I may just hack that in myself in a subsequent PR.~~ (see EDIT 2, below)

### Details

- The formatting style is defined by the config file `components/eamxx/.clang-format`.
  - I've chosen the [LLVM style](https://llvm.org/docs/CodingStandards.html) for formatting, since that is what we've been using in MAM4xx and seems reasonably non-controversial.
  - I have not added any customizations to the `.clang-format` config, but I have included, as comments, all of the options that the LLVM style imposes.
  - I expect there will be opinions on formatting choices, so please reach out with comments.
- Different `clang` versions have ***slightly*** different behaviors, so I've chosen `clang-format` v14, again, because that's what we use in MAM4xx.
- This is intended to be low-impact on developers, so please reach out with any questions, comments, angry rants, etc. 🙂

#### EDIT

I've started a [discussion page](https://github.com/E3SM-Project/E3SM/discussions/7307) where others can provide feedback.

#### EDIT 2

As mentioned above, I modified the `DoozyX/clang-format-lint-action` Action so that it prints a step summary indicating which, if any, files fail the `clang-format` test. The fork of that Action is [here](https://github.com/mjschmidt271/clang-format-lint-action)

---

# DRAFT - PR Must be Open to Trigger Testing

This new GitHub Action is intended to do a formatting check **only on the files changed by the PR** using the `clang-format` tool.

To begin with, I am testing 2 different actions provided by the GitHub Marketplace. One of them only returns a binary pass/fail, and the other can change the files in-place (not the desired behavior) but it's not clear to me whether disabling that option returns a diff and whether it writes new files with the changes or merely outputs the diff to the log.

bartgol commented on .github/workflows/e3sm-gh-clang-format.yml:
```diff
@@ -0,0 +1,73 @@
+    paths:
+      - 'components/eamxx/**/*.cpp'
+      - 'components/eamxx/**/*.hpp'
+      - '.github/workflows/e3sm-gh-clang-format.yml'
```
Should we call the file `eamxx-gh-clang-format.yml`, to avoid giving false expectations?

bartgol commented on components/eamxx/src/atmosphere_driver.hpp:
```diff
@@ -0,0 +1,264 @@
+#ifndef SCREAM_ATMOSPHERE_DRIVER_HPP
```
I'm assuming you did not mean to create these copies of the AD in the src folder?

bartgol commented on components/eamxx/.clang-format:
```diff
@@ -1,15 +1,196 @@
-UseTab:            Never
-TabWidth:          2
-ColumnLimit:       80
+BasedOnStyle: LLVM
```
I thought we agreed that each component can format whichever way they want.

bartgol replied:
I'm ok keeping the commented options for a week or two, to give ppl the chance to try out some options just by uncommenting lines, and then sharing thoughts in the discussion page.

But I don't feel strongly about it, so if you want to rm commented lines, go ahead.

rljacob commented:
This will check for format but not automatically change the file?

bartgol commented:
> This will check for format but not automatically change the file?

That's my understanding. I think the idea would be that once the PR is good to merge, the author (or maybe the integrator) runs the clang formatter, formats the files, pushes a new commit and then merges. I don't want the action to add a commit when it runs, since if the PR is open for a while, it may run several times, and commit multiple times a "clang format" commit.

jgfouca commented:
This doesn't force us into <=80 line length, right? That would be super annoying and makes the code look like crap (IMO).

philipwjones commented:
We restrict to 80 char since that is a typical default terminal/editor window and for many of us, the optimal visual range. Having said that, it's definitely better to make those line breaks yourself in logical places, so even if the formatter catches it and tries to do it for you, you might want to manually edit those to get a better line break.

philipwjones commented:
Also, 80-char limit is part of the LLVM coding style

bartgol commented:
I am going to push back on the 80 chars limit. This is one of the reasons I never pushed for a formatter: everyone likes different lengths, tabs, indentation rules, etc.

Even though 80 chars are common, I find the width too narrow. Yes, one can break long templates/signatures over multiple lines, but sometimes i prefer going a few chars long than breaking the line. And with multiple indentation levels (namespace, class, loop, lambda, nested loops, etc) one is often starting with a few chars already gone.

Not just that, I fear someone may start using abbrev nms just 2 avoid line ovrflw... I really like when we use long, verbose names, but I would like to avoid having every line of code broken over 2-3 lines...

If we really REALLY want a limit, I would request at least 100 (possibly 120?).

jgfouca commented:
Yeah, I went through this last summer with KokkosKernels. Extremely complex code with heavy templating just looks like crap with an 80 character limit as the formatter chops and squeezes your code. We (KK) bumped it to 120 and it's so much nicer.

mjschmidt271 commented on .github/workflows/eamxx-gh-clang-format.yml:
```diff
@@ -0,0 +1,36 @@
+        files: 'components/eamxx/**/*.{cpp,hpp}'
+        separator: " "
+    # displays the diff, comparing to the changes that *would* be made
+    - uses: DoozyX/clang-format-lint-action@v0.20
```
We certainly could. Having given it a close look while modifying the `DoozyX` action, it's reasonably straightforward and I feel comfortable debugging any future issues. Currently, if a file fails the `clang-format` the action prints a diff to the log between the original and the changes that would bring it into compliance. It then prints a Step Summary to the workflow page listing any files that fail.

When you say "exact trigger" do you mean the function signature/arguments for the `clang-format` call? I also modified the comments at this place in the code to try and make things clearer

mam4xxSNL commented:
Alright, I do believe I've addressed all of the comments. As for style conventions, here's where we're at:

Finally, I would ask to hold off on more comments on the conventions until this week's EAMxx dev meeting, and I'll make sure this topic is on the agenda!

@mahf708 @bartgol @jgfouca @rljacob

mjschmidt271 commented:
Well, that's strange... switching GH accounts with a partiallly-written message can apparently send it(?)? Anyway...

Alright, I do believe I've addressed all of the comments. As for style conventions, here's where we're at:

- BasedOnStyle: LLVM
- ColumnLimit: 100
- AlignConsecutiveAssignments: true
- AlignConsecutiveBitFields: true
- AlignConsecutiveMacros: true
- AlignEscapedNewlines: true
- AlignTrailingComments: true

This is based on requests to expand the 80 column limit for lines, under the reasoning to try out 100, rather than going for 120, as some have mentioned.
Based on feedback from the Omega team, I'd advocate for trying out the alignment settings.
And there's not been much action on other formatting topics, so we will keep them default for now.
In 2 weeks (5/20), I'll remove the commented lines from the [.clang-format](components/eamxx/.clang-format) config.

Finally, I would ask to hold off on more comments on the conventions until this week's EAMxx dev meeting, and I'll make sure this topic is on the agenda!

@mahf708 @bartgol @jgfouca @rljacob

bartgol commented on .github/workflows/eamxx-gh-clang-format.yml:
```diff
@@ -0,0 +1,42 @@
+      with:
+        files: 'components/eamxx/**/*.{cpp,hpp}'
+        separator: " "
+    # 1. runs clang-format on the files changed by this PR, and returns pass/fail
+    #    error code.
+    # 2. prints the diff to screen (action log), comparing to the changes
+    #    that *would* be made
+    # 3. the mjschmidt fork of DoozyX/clang-format-lint-action@v0.20
+    #    adds a Step summary to the workflow page, and, on failure, lists the
+    #    files that fail the clang-format test
+    - uses: mjschmidt271/clang-format-lint-action@v1.0.0
+      with:
+        source: ${{ steps.changed-files.outputs.all_changed_files }}
+        exclude: ''
+        extensions: 'hpp,cpp'
+        clangFormatVersion: 14
+        style: 'file:components/eamxx/.clang-format'
```
Personally, I find composite actions nicer than `run: blah` steps.

That said, I think e3sm's mantra is to avoid stuff from personal repos. Any interest in forking the clang-format-lint action in e3sm-project?

mjschmidt271 replied:
@bartgol - I do agree that forking the action to `e3sm-project` would make the most sense. I'll reach out elsewhere about that.

And, having now chatted with @mahf708 about this, I plan to include the `clang-format [...]` invocation signature in the log output and also the job summary

bartgol commented:
@mjschmidt271 It appears the action repo is still in your fork. Also, you added back the driver files (I assume to test the action would work, and planned to remove later?)

mjschmidt271 commented:
> @mjschmidt271 It appears the action repo is still in your fork. Also, you added back the driver files (I assume to test the action would work, and planned to remove later?)

@bartgol Exactly correct--I've now switched to the `E3SM-Project` fork, removed files, and rebased. Should be ready to go!
